### PR TITLE
Refactor CI workflows to leverage matrices for less code duplication

### LINKED ISSUE #1701: Leverage matrixes in GitHub CI jobs to reduce code duplication
### Is your feature request related to a specific problem?

There are lots of instances of code duplication in our Workflow files that could get refactored away with proper usage of matrixes

### PR DESCRIPTION
**Pull request by ThomasLaPiana**
Closes #1701 

### Code Changes

* [x] update the static checks to use a matrix
* [x] refactor the "safe tests" to use a matrix
* [x] refactor the "fides checks" into the safe tests/static checks

### Steps to Confirm

* [ ] look through all of the CI tests running on this PR and make sure that 1) the jobs are correct and 2) the amount of jobs is equal to the number on other PRs, etc. 

*NOTE*: The number of jobs should be 35 total, previously it was 33 but we separated the Fides checks into 3 separate jobs so now there are 2 extra

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation Updated:
  * [x] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  * [ ] documentation issue created (tag docs-team to complete issue separately)
* [x] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created
* [ ] ~~Update `CHANGELOG.md`~~

### Description Of Changes

There is a lot of unnecessary code duplication in our CI jobs, mostly repetitive setup steps where the only thing changing is the final command. This can be done in a smarter way via usage of matrixes

The unsafe tests were _not_ updated to use matrices due to concerns about race conditions for external tests, and was noted as such in the workflow file

### CODE REVIEW

**Review by sanders41 — CHANGES_REQUESTED** [review]

File: .github/workflows/backend_checks.yml [hunk 1]
```diff
@@ -44,127 +45,65 @@ jobs:
...
-        uses: actions/checkout@v3
+        test_selection:
+        - "pytest_ctl(not-external)"
+        - "pytest_ops(unit)"
+        - "pytest_ops(integration)"
+        - "check_fides_annotations"
+        - "fides_db_scan"
+        - "docs_checks"
+        exclude:
+        # Not all jobs need to run against the Python matrix
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply .github/workflows/backend_checks.yml:91]: @sanders41 they're tests/checks that are unrelated to Python version, as they are testing some part of the application/build that we control and always run consistently ourselves, nothing here is user-facing (i.e. our own docs builds, our own fides annotations, etc.)

File: .github/workflows/backend_checks.yml [hunk 2]
```diff
@@ -44,127 +45,65 @@ jobs:
...
-          name: python-${{ matrix.python_version }}
-          path: /tmp/
-
-      - name: Load image
-        run: docker load --input /tmp/python-${{ matrix.python_version }}.tar
-
-      - name: Checkout
-        uses: actions/checkout@v3
+        test_selection:
+        - "pytest_ctl(not-external)"
```
[the remark this answers is not in the document]
Comment (ThomasLaPiana) [reply .github/workflows/backend_checks.yml:84]: good call, should be consistent. Guess my formatter isn't working!

**Comment by ThomasLaPiana** [comment]
@sanders41 thanks for the early review comments! good suggestions

**Comment by ThomasLaPiana** [comment]
Current failures are unrelated:

Ctl tests are fixed in #1700 
Ops tests are due to Shopify API limits, and are unrelated as well #1714

**Comment by ThomasLaPiana** [comment]
Current failures are unrelated:

Ctl tests are fixed in #1700 
Ops tests are due to Shopify API limits, and are unrelated as well #1714

**Pull request merged**